### DISCUSSION TITLE
Enabling the `quarkus.micrometer.binder.kafka.enabled` property

### DISCUSSION THREAD

**Discussion by ShubhamRwt** [quarkusio/quarkus#32639] [Q&A]
Hello,

I am working on moving the Strimzi Kafka bridge project with @ppatierno  to use Quarkus and the HTTP REST part is already implemented using RESTeasy Reactive.
I am currently trying to add the metrics to bridge using the `quarkus micrometer` library. I was able to get the metrics which are there by default on http://localhost:8080/q/metrics. I saw that using the `quarkus.micrometer.binder.kafka.enabled` we can enable the Kafka related metrics but I am not able to see any metrics related to Kafka after setting this property to `true`. I also fulfil the conditions  mentioned in [the docs](https://quarkus.io/guides/micrometer#quarkus-micrometer_quarkus.micrometer.binder.kafka.enabled) but still no good. Can you help me figure out  where I am going wrong or why it is not working?

[a comment by an automated account is not shown]

**Reply by ppatierno** [reply]
Thanks @ShubhamRwt for opening this!
Just to add that we are not using Kafka through the SmallRye dependency but directly the Kafka Java client through:

```xml
<dependency>
      <groupId>org.apache.kafka</groupId>
      <artifactId>kafka-clients</artifactId>
      <version>${kafka.version}</version>
</dependency>
```

it should fulfil the condition mentioned by the doc about `Kafka Consumer or Producer interface is on the classpath`

**Reply by brunobat** [reply]
That's correct.
Also, it applies only to kafka-clients, not kafka-streams. See: https://github.com/quarkusio/quarkus/blob/main/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/kafka/KafkaEventObserver.java#L21

**Reply by ShubhamRwt** [reply]
But we are still not getting any metrics related to kafka. Even when the conditions are fulfilled :(

**Comment by alesj** [comment]
@ShubhamRwt can you debug and see if this observer kicks-in?

**Reply by ppatierno** [reply]
@alesj so we have this dependency for micrometer ...

```xml
<dependency>
     <groupId>io.quarkus</groupId>
     <artifactId>quarkus-micrometer-registry-prometheus</artifactId>
</dependency>
```

then `quarkus.micrometer.binder.kafka.enabled=true` in the application.properties.

When our code executes this kind of code for example (producer creation):

```java
this.producer = new KafkaProducer<>(props, this.keySerializer, this.valueSerializer);
```

no breakpoint hits in the `KafkaEventObserver.producerCreated`

**Reply by alesj** [reply]
OK, I think we established what was the issue -- producers and consumers were not beans.
Let me know when the KafkaEvenObserver injection idea works for you, and we can close the issue.

**Reply by ppatierno** [reply]
Yes it worked. We inject the `KafkaEvenObserver` instance where we can and pass it down to the chain where we need it (which is not a bean).
Not sure we are going to use it just because our way to collect the metrics (via JMX collector) is anyway needed for getting others (not from Kafka client itself).
Anyway, I would say that this feature is really super useful :-)
Out of curiosity, how much is it customizable? i.e. adding a prefix to the metrics name, adding specific tags, filtering, and so on?

**Comment by alesj** [comment]
> Out of curiosity, how much is it customizable? i.e. adding a prefix to the metrics name, adding specific tags, filtering, and so on?

This is out of Quarkus' "jurisdiction", as this is then Micrometer + Kafka code.
* see `KafkaMetrics` class in micrometer-core artifact

**Comment by alesj** [comment]
@ppatierno it takes Metrics from consumer/producer/admin ... and turns them into Micrometer metrics.
I guess if you can modify the metrics (before you pass them to KafkaEventObserver) on the consumer/producer/admin, it should reflect in the Micrometer metrics.

**Reply by ebullient** [reply]
Micrometer also has MeterFilter, which you can use to alter metric names and tags (though there are constraints on the information available if you want to add tags).

https://quarkus.io/guides/micrometer#using-meterfilter-to-configure-metrics

Remember that metrics are for use in aggregate as indicators, not for audit/debug purposes. ;)

**Reply by ppatierno** [reply]
> Remember that metrics are for use in aggregate as indicators, not for audit/debug purposes. ;)

Wdym by this? We don't want to use metrics for audit/debug purposes.

**Reply by ebullient** [reply]
I've seen folks start putting things in tags that are too finely detailed... so, was just a reminder. ;)
